File: crates/scp2p-desktop/src/commands.rs
// Copyright (c) 2024-2026 Vanyo Vanev / Tech Art Ltd
// SPDX-License-Identifier: MPL-2.0
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
use crate::{
    app_state::DesktopAppState,
    dto::{
        CommunityBrowseView, CommunityView, DesktopClientConfig, OwnedShareView, PeerView,
        PublicShareView, PublishResultView, PublishVisibility, RuntimeStatus, SearchResultsView,
        ShareItemView, StartNodeRequest, SubscriptionView,
    },
};

pub async fn start_node(
    app_state: &DesktopAppState,
    request: StartNodeRequest,
) -> anyhow::Result<RuntimeStatus> {
    app_state.start_node(request).await
}

pub async fn stop_node(app_state: &DesktopAppState) -> anyhow::Result<RuntimeStatus> {
    Ok(app_state.stop_node().await)
}

pub async fn runtime_status(app_state: &DesktopAppState) -> anyhow::Result<RuntimeStatus> {
    app_state.status().await
}

pub async fn save_client_config(
    app_state: &DesktopAppState,
    path: String,
    config: DesktopClientConfig,
) -> anyhow::Result<()> {
    app_state.save_client_config(path, &config).await
}

pub async fn load_client_config(
    app_state: &DesktopAppState,
    path: String,
) -> anyhow::Result<DesktopClientConfig> {
    app_state.load_client_config(path).await
}

pub async fn list_peers(app_state: &DesktopAppState) -> anyhow::Result<Vec<PeerView>> {
    app_state.peer_views().await
}

pub async fn list_subscriptions(
    app_state: &DesktopAppState,
) -> anyhow::Result<Vec<SubscriptionView>> {
    app_state.subscription_views().await
}

pub async fn list_communities(app_state: &DesktopAppState) -> anyhow::Result<Vec<CommunityView>> {
    app_state.community_views().await
}

pub async fn join_community(
    app_state: &DesktopAppState,
    share_id_hex: String,
    share_pubkey_hex: String,
) -> anyhow::Result<Vec<CommunityView>> {
    app_state
        .join_community(&share_id_hex, &share_pubkey_hex)
        .await
}

pub async fn browse_community(
    app_state: &DesktopAppState,
    share_id_hex: String,
) -> anyhow::Result<CommunityBrowseView> {
    app_state.browse_community(&share_id_hex).await
}

pub async fn subscribe_share(
    app_state: &DesktopAppState,
    share_id_hex: String,
) -> anyhow::Result<Vec<SubscriptionView>> {
    app_state.subscribe_share(&share_id_hex).await
}

pub async fn unsubscribe_share(
    app_state: &DesktopAppState,
    share_id_hex: String,
) -> anyhow::Result<Vec<SubscriptionView>> {
    app_state.unsubscribe_share(&share_id_hex).await
}

pub async fn sync_now(app_state: &DesktopAppState) -> anyhow::Result<Vec<SubscriptionView>> {
    app_state.sync_now().await
}

pub async fn search_catalogs(
    app_state: &DesktopAppState,
    text: String,
) -> anyhow::Result<SearchResultsView> {
    app_state.search_catalogs(&text).await
}

pub async fn browse_public_shares(
    app_state: &DesktopAppState,
) -> anyhow::Result<Vec<PublicShareView>> {
    app_state.browse_public_shares().await
}

pub async fn subscribe_public_share(
    app_state: &DesktopAppState,
    one_based_index: usize,
) -> anyhow::Result<Vec<SubscriptionView>> {
    app_state.subscribe_public_share(one_based_index).await
}

pub async fn download_content(
    app_state: &DesktopAppState,
    content_id_hex: String,
    target_path: String,
) -> anyhow::Result<()> {
    app_state
        .download_content(&content_id_hex, &target_path)
        .await
}

pub async fn publish_text_share(
    app_state: &DesktopAppState,
    title: String,
    item_name: String,
    item_text: String,
    visibility: PublishVisibility,
    community_ids_hex: Vec<String>,
) -> anyhow::Result<PublishResultView> {
    app_state
        .publish_text_share(
            &title,
            &item_name,
            &item_text,
            visibility,
            &community_ids_hex,
        )
        .await
}

pub async fn publish_files(
    app_state: &DesktopAppState,
    file_paths: Vec<String>,
    title: String,
    visibility: PublishVisibility,
    community_ids_hex: Vec<String>,
) -> anyhow::Result<PublishResultView> {
    app_state
        .publish_files(&file_paths, &title, visibility, &community_ids_hex)
        .await
}

pub async fn publish_folder(
    app_state: &DesktopAppState,
    dir_path: String,
    title: String,
    visibility: PublishVisibility,
    community_ids_hex: Vec<String>,
) -> anyhow::Result<PublishResultView> {
    app_state
        .publish_folder(&dir_path, &title, visibility, &community_ids_hex)
        .await
}

pub async fn browse_share_items(
    app_state: &DesktopAppState,
    share_id_hex: String,
) -> anyhow::Result<Vec<ShareItemView>> {
    app_state.browse_share_items(&share_id_hex).await
}

pub async fn download_share_items(
    app_state: &DesktopAppState,
    share_id_hex: String,
    content_ids_hex: Vec<String>,
    target_dir: String,
    on_progress: Option<&scp2p_core::ProgressCallback>,
) -> anyhow::Result<Vec<String>> {
    app_state
        .download_share_items(&share_id_hex, &content_ids_hex, &target_dir, on_progress)
        .await
}

// ── My Shares ──────────────────────────────────────────────────────────────

pub async fn list_my_shares(
    app_state: &DesktopAppState,
) -> anyhow::Result<Vec<OwnedShareView>> {
    app_state.list_my_shares().await
}

pub async fn delete_my_share(
    app_state: &DesktopAppState,
    share_id_hex: String,
) -> anyhow::Result<Vec<OwnedShareView>> {
    app_state.delete_my_share(&share_id_hex).await
}

pub async fn update_my_share_visibility(
    app_state: &DesktopAppState,
    share_id_hex: String,
    visibility: PublishVisibility,
) -> anyhow::Result<Vec<OwnedShareView>> {
    app_state
        .update_my_share_visibility(&share_id_hex, visibility)
        .await
}

File: app/src/lib/commands.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  RuntimeStatus,
  StartNodeRequest,
  DesktopClientConfig,
  PeerView,
  SubscriptionView,
  CommunityView,
  CommunityBrowseView,
  SearchResultsView,
  PublicShareView,
  PublishResultView,
  PublishVisibility,
  ShareItemView,
  OwnedShareView,
} from "./types";

// ── Node lifecycle ──────────────────────────────────────────────────────

export async function startNode(
  request: StartNodeRequest
): Promise<RuntimeStatus> {
  return invoke("start_node", { request });
}

export async function stopNode(): Promise<RuntimeStatus> {
  return invoke("stop_node");
}

export async function runtimeStatus(): Promise<RuntimeStatus> {
  return invoke("runtime_status");
}

// ── Config ──────────────────────────────────────────────────────────────

export async function saveClientConfig(
  path: string,
  config: DesktopClientConfig
): Promise<void> {
  return invoke("save_client_config", { path, config });
}

export async function loadClientConfig(
  path: string
): Promise<DesktopClientConfig> {
  return invoke("load_client_config", { path });
}

export async function autoStartNode(
  configPath: string
): Promise<RuntimeStatus | null> {
  return invoke("auto_start_node", { configPath });
}

// ── Peers ───────────────────────────────────────────────────────────────

export async function listPeers(): Promise<PeerView[]> {
  return invoke("list_peers");
}

// ── Subscriptions ───────────────────────────────────────────────────────

export async function listSubscriptions(): Promise<SubscriptionView[]> {
  return invoke("list_subscriptions");
}

export async function subscribeShare(
  shareIdHex: string
): Promise<SubscriptionView[]> {
  return invoke("subscribe_share", { shareIdHex });
}

export async function unsubscribeShare(
  shareIdHex: string
): Promise<SubscriptionView[]> {
  return invoke("unsubscribe_share", { shareIdHex });
}

export async function syncNow(): Promise<SubscriptionView[]> {
  return invoke("sync_now");
}

// ── Communities ─────────────────────────────────────────────────────────

export async function listCommunities(): Promise<CommunityView[]> {
  return invoke("list_communities");
}

export async function joinCommunity(
  shareIdHex: string,
  sharePubkeyHex: string
): Promise<CommunityView[]> {
  return invoke("join_community", { shareIdHex, sharePubkeyHex });
}

export async function leaveCommunity(
  shareIdHex: string
): Promise<CommunityView[]> {
  return invoke("leave_community", { shareIdHex });
}

export async function browseCommunity(
  shareIdHex: string
): Promise<CommunityBrowseView> {
  return invoke("browse_community", { shareIdHex });
}

// ── Search ──────────────────────────────────────────────────────────────

export async function searchCatalogs(
  text: string
): Promise<SearchResultsView> {
  return invoke("search_catalogs", { text });
}

// ── Public shares ───────────────────────────────────────────────────────

export async function browsePublicShares(): Promise<PublicShareView[]> {
  return invoke("browse_public_shares");
}

export async function subscribePublicShare(
  oneBasedIndex: number
): Promise<SubscriptionView[]> {
  return invoke("subscribe_public_share", { oneBasedIndex });
}

// ── Download ────────────────────────────────────────────────────────────

export async function downloadContent(
  contentIdHex: string,
  targetPath: string
): Promise<void> {
  return invoke("download_content", { contentIdHex, targetPath });
}

// ── Publish ─────────────────────────────────────────────────────────────

export async function publishTextShare(
  title: string,
  itemName: string,
  itemText: string,
  visibility: PublishVisibility,
  communityIdsHex: string[]
): Promise<PublishResultView> {
  return invoke("publish_text_share", {
    title,
    itemName,
    itemText,
    visibility,
    communityIdsHex,
  });
}

export async function publishFiles(
  filePaths: string[],
  title: string,
  visibility: PublishVisibility,
  communityIdsHex: string[]
): Promise<PublishResultView> {
  return invoke("publish_files", {
    filePaths,
    title,
    visibility,
    communityIdsHex,
  });
}

export async function publishFolder(
  dirPath: string,
  title: string,
  visibility: PublishVisibility,
  communityIdsHex: string[]
): Promise<PublishResultView> {
  return invoke("publish_folder", {
    dirPath,
    title,
    visibility,
    communityIdsHex,
  });
}

// ── Share Browser ───────────────────────────────────────────────────────

export async function browseShareItems(
  shareIdHex: string
): Promise<ShareItemView[]> {
  return invoke("browse_share_items", { shareIdHex });
}

export async function downloadShareItems(
  shareIdHex: string,
  contentIdsHex: string[],
  targetDir: string
): Promise<string[]> {
  return invoke("download_share_items", {
    shareIdHex,
    contentIdsHex,
    targetDir,
  });
}

// ── My Shares ───────────────────────────────────────────────────────────

export async function listMyShares(): Promise<OwnedShareView[]> {
  return invoke("list_my_shares");
}

export async function deleteMyShare(
  shareIdHex: string
): Promise<OwnedShareView[]> {
  return invoke("delete_my_share", { shareIdHex });
}

export async function updateMyShareVisibility(
  shareIdHex: string,
  visibility: PublishVisibility
): Promise<OwnedShareView[]> {
  return invoke("update_my_share_visibility", { shareIdHex, visibility });
}
